src/hotspot/share/ci/ciMethodBlocks.cpp

```diff
@@ -158,10 +158,13 @@ void ciMethodBlocks::do_analysis() {
       case Bytecodes::_goto        :
       {
         cur_block->set_control_bci(bci);
+        int dest_bci = s.get_dest();
         if (s.next_bci() < limit_bci) {
-          (void) make_block_at(s.next_bci());
+          ciBlock *next_block = make_block_at(s.next_bci());
+          if (dest_bci < bci) {
+            next_block->set_is_loop_exit();
+          }
```

Comment:
I think this loop detection logic is both wrong and incomplete. For example, javac generates no `goto` for the following loop:
```
        int i = 0;
        do {
            
        } while (i++ < 10);
```

And in the following case, the block after the `goto` corresponding to the `continue` statement is not the loop exit:
```
        int i = 0;
        label:
        while (true) {
            i++;
            if (i == 1)
                continue label;
            if (i == 2)
                break;
        }
```

I just quickly hacked this, there are probably better examples.

Reply:
Nice catch!

I missed the `do {...} while (condition)` loop pattern and the `continue` statement.
So we should find better way to identify the loop exit block, right?